spring-integration-test/src/main/java/org/springframework/integration/test/mock/MockMessageHandler.java

```diff
@@ -26,19 +24,39 @@
 import org.hamcrest.Matcher;
 
 import org.springframework.integration.handler.AbstractMessageProducingHandler;
+import org.springframework.integration.test.matcher.PayloadAndHeaderMatcher;
+import org.springframework.integration.test.matcher.PayloadMatcher;
 import org.springframework.messaging.Message;
+import org.springframework.util.Assert;
 
 /**
+ * The {@link AbstractMessageProducingHandler} extension for the assertion purpose in tests.
+ * <p>
+ * The provided {@link Matcher}s are applied to the incoming messages one at a time
+ * until the last {@link Matcher}, which is applied for all subsequent messages -
+ * the similar behavior exists in the
+ * {@code Mockito.doReturn(Object toBeReturned, Object... toBeReturnedNext)}.
+ * <p>
+ * Typically is used as a chain of assertions and optional replies for them:
+ * <pre class="code">
+ * {@code
+ *      MockIntegration.mockMessageHandler(hasHeader("bar", "BAR"))
+ *               .thenReply()
+ *               .assertNext(new GenericMessage<>("foo", Collections.singletonMap("key", "value")))
+ *               .assertNext(hasPayload("foo"))
```

Comment:
I am a little torn here; I tend to agree to @Toparvion that it's a bit confusing but I can it's useful, especially with `thenReply(m -> ...)` since you will likely generate the replies based on the request.

I am not sure there's much value in `thenReply()` echoing back the same message so I wouldn't mind it being removed altogether - it's name is a bit blah too (see comment elsewhere in this review).

I also don't really like the first matcher being the factory method argument.

How about...

    MockIntegration.mockMessageHandler()
             .replyFunction(m -> ...)
             .assertNext(...)
             .assertNext(...)
             .assertNext(...).withReply(m -> ...)
             .assertNext(...)

or...

    MockIntegration.mockMessageHandler()
             .replyFunction(m -> ...)
             .assertNext(...)
             .assertNext(...)
             .replyFunction(m -> ...)
             .assertNext(...)
             .assertNext(...)